File: src/main/test/net/jcflorezr/audiocontent/signal/SoundZonesDetectorImplTest.java
package net.jcflorezr.audiocontent.signal;

import biz.source_code.dsp.model.AudioSignal;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import net.jcflorezr.model.audioclips.AudioFileClipEntity;
import net.jcflorezr.model.audiocontent.signal.RmsSignalInfo;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.List;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anyObject;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class SoundZonesDetectorImplTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final String BACKGROUND_NOISE_LOW_VOLUME_SIGNAL = "/audiocontent/signal/background-noise-low-volume-signal.json";
    private static final String BACKGROUND_NOISE_LOW_VOLUME_RMS_INFO = "/audiocontent/signal/background-noise-low-volume-rms-info.json";
    private static final String BACKGROUND_NOISE_LOW_VOLUME_AUDIO_CLIPS_INFO = "/audiocontent/signal/background-noise-low-volume-audio-clips-info.json";
    private static final String WITH_APPLAUSE_SIGNAL = "/audiocontent/signal/with-applause-signal.json";
    private static final String WITH_APPLAUSE_RMS_INFO = "/audiocontent/signal/with-applause-rms-info.json";
    private static final String WITH_APPLAUSE_AUDIO_CLIPS_INFO = "/audiocontent/signal/with-applause-audio-clips-info.json";
    private static final String STRONG_BACKGROUND_NOISE_SIGNAL = "/audiocontent/signal/strong-background-noise-signal.json";
    private static final String STRONG_BACKGROUND_NOISE_RMS_INFO = "/audiocontent/signal/strong-background-noise-rms-info.json";
    private static final String STRONG_BACKGROUND_NOISE_AUDIO_CLIPS_INFO = "/audiocontent/signal/strong-background-noise-audio-clips-info.json";

    private Class<? extends SoundZonesDetectorImplTest> thisClass;

    @Mock
    private RmsCalculator rmsCalculator;
    @InjectMocks
    private SoundZonesDetectorImpl soundZonesDetector;

    @Before
    public void setUp() throws Exception {
        thisClass = this.getClass();
    }

    @Test
    public void retrieveAudioClipsWithBackgroundNoiseAndLowVolume() throws Exception {
        testAudioClipsProcess(BACKGROUND_NOISE_LOW_VOLUME_RMS_INFO,
                              BACKGROUND_NOISE_LOW_VOLUME_SIGNAL,
                              BACKGROUND_NOISE_LOW_VOLUME_AUDIO_CLIPS_INFO);
    }

    @Test
    public void retrieveAudioClipsWithApplause() throws Exception {
        testAudioClipsProcess(WITH_APPLAUSE_RMS_INFO,
                              WITH_APPLAUSE_SIGNAL,
                              WITH_APPLAUSE_AUDIO_CLIPS_INFO);
    }

    @Test
    public void retrieveAudioClipsWithStrongBackgroundNoise() throws Exception {
        testAudioClipsProcess(STRONG_BACKGROUND_NOISE_RMS_INFO,
                              STRONG_BACKGROUND_NOISE_SIGNAL,
                              STRONG_BACKGROUND_NOISE_AUDIO_CLIPS_INFO);
    }

    private void testAudioClipsProcess(String rmsSignalInfoFileName, String audioSignalFileName, String expectedAudioClipsFileName) throws Exception {
        List<RmsSignalInfo> rmsSignalInfo = MAPPER.readValue(thisClass.getResourceAsStream(rmsSignalInfoFileName), new TypeReference<List<RmsSignalInfo>>() {});
        when(rmsCalculator.retrieveRmsInfo(anyObject(), anyInt(), anyInt())).thenReturn(rmsSignalInfo);

        String audioFileName = "/path/to-find/audio-file";
        AudioSignal withApplauseSignal = MAPPER.readValue(thisClass.getResourceAsStream(audioSignalFileName), AudioSignal.class);
        List<AudioFileClipEntity> actualAudioClips = soundZonesDetector.retrieveAudioClipsInfo(audioFileName, withApplauseSignal);
        List<AudioFileClipEntity> expectedAudioClips = MAPPER.readValue(thisClass.getResourceAsStream(expectedAudioClipsFileName), new TypeReference<List<AudioFileClipEntity>>() {});

        assertThat(actualAudioClips, equalTo(expectedAudioClips));
    }

}

File: src/test/kotlin/net/jcflorezr/signal/SoundZonesDetectorImplTest.kt
package net.jcflorezr.signal

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import net.jcflorezr.config.TestRootConfig
import net.jcflorezr.model.AudioClipInfo
import net.jcflorezr.model.AudioSignalRmsInfoKt
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit4.SpringJUnit4ClassRunner
import java.io.File
import org.hamcrest.CoreMatchers.`is` as Is

@ActiveProfiles("test")
@RunWith(SpringJUnit4ClassRunner::class)
@ContextConfiguration(classes = [TestRootConfig::class])
class SoundZonesDetectorImplTest {

    @Autowired
    private lateinit var soundZonesDetector: SoundZonesDetector

    companion object {
        private val MAPPER = ObjectMapper().registerKotlinModule()
    }

    private val signalResourcesPath: String
    private val clipsResourcesPath: String
    private val thisClass: Class<SoundZonesDetectorImplTest> = this.javaClass

    init {
        signalResourcesPath = thisClass.getResource("/signal/").path
        clipsResourcesPath = thisClass.getResource("/clip/").path
    }

    @Test
    fun generateAudioClipsInfoForFileWithBackgroundNoiseAndLowVoiceVolume() {
        detectSoundZones(
            path = signalResourcesPath + "background-noise-low-volume/background-noise-low-volume.json",
            folderName = "background-noise-low-volume"
        )
    }

//    @Test
//    fun generateIncompleteAudioClipsInfoForFileWithBackgroundNoiseAndLowVoiceVolume() {
//        detectSoundZones(
//            path = signalResourcesPath + "background-noise-low-volume/background-noise-low-volume-incomplete.json",
//            folderName = "background-noise-low-volume"
//        )
//    }

    @Test
    fun generateAudioClipsInfoForFileWithApplause() {
        detectSoundZones(
            path = signalResourcesPath + "with-applause/with-applause.json",
            folderName = "with-applause"
        )
    }

//    @Test
//    fun generateIncompleteAudioClipsInfoForFileWithApplause() {
//        detectSoundZones(
//            path = signalResourcesPath + "with-applause/with-applause-incomplete.json",
//            folderName = "with-applause"
//        )
//    }

    @Test
    fun generateAudioClipsInfoForFileWithStrongBackgroundNoise() {
        detectSoundZones(
            path = signalResourcesPath + "strong-background-noise/strong-background-noise.json",
            folderName = "strong-background-noise"
        )
    }

    private fun detectSoundZones(path: String, folderName: String) {
//        val signalRmsListType = MAPPER.typeFactory.constructCollectionType(List::class.java, AudioSignalRmsInfoKt::class.java)
//        val audioSignalRmsList: List<AudioSignalRmsInfoKt> = MAPPER.readValue(File(path), signalRmsListType)
//        val actualAudioClipsInfoList = soundZonesDetector.generateSoundZones(audioRmsInfoList = audioSignalRmsList)
//        val audioClipListType = MAPPER.typeFactory.constructCollectionType(List::class.java, AudioClipInfo::class.java)
//        val expectedAudioClipsInfoList: List<AudioClipInfo> =
//            MAPPER.readValue(File("$clipsResourcesPath/$folderName/$folderName.json"), audioClipListType)
//        assertThat(actualAudioClipsInfoList, Is(equalTo(expectedAudioClipsInfoList)))
    }
}